repose-aggregator/components/filters/saml-policy-translation-filter/src/main/scala/org/openrepose/filters/samlpolicy/SamlIdentityClient.scala

```diff
@@ -0,0 +1,262 @@
+          .getOrElse(akkaServiceClientFactory.newAkkaServiceClient)
+      )
+    }
+
+    val optPolicyServiceClient = Option(policyServiceClient)
+    if (optPolicyServiceClient.isEmpty || policyServiceClient.poolId != policyPoolId) {
+      optPolicyServiceClient.foreach(_.akkaServiceClient.destroy())
+      policyServiceClient = ServiceClient(
+        policyPoolId,
+        policyPoolId.map(akkaServiceClientFactory.newAkkaServiceClient)
+          .getOrElse(akkaServiceClientFactory.newAkkaServiceClient)
+      )
+    }
+  }
+
+  /**
+    * Retrieves a token from Identity that will be used for authorization on all other calls to Identity.
+    *
+    * @param username the username on the account
+    * @param password the password on the account
+    * @param traceId an optional identifier to be sent with the request
+    * @return the token if successful, or a failure if unsuccessful
+    */
+  def getToken(username: String, password: String, traceId: Option[String]): Try[String] = {
+    val authenticationPayload = Json.obj(
+      "auth" -> Json.obj(
+        "passwordCredentials" -> Json.obj(
+          "username" -> username,
+          "password" -> password
+        )
+      )
+    )
+
+    val akkaResponse = Try(tokenServiceClient.akkaServiceClient.post(
+      TokenRequestKey,
+      s"$tokenUri$TokenPath",
+      (Map(CommonHttpHeader.ACCEPT.toString -> MediaType.APPLICATION_JSON)
+        ++ traceId.map(CommonHttpHeader.TRACE_GUID.toString.->)).asJava,
+      Json.stringify(authenticationPayload),
+      MediaType.APPLICATION_JSON_TYPE
+    ))
+
+    akkaResponse match {
+      case Success(serviceClientResponse) =>
+        serviceClientResponse.getStatus match {
+          case SC_OK =>
+            Try {
+              val responseEncoding = serviceClientResponse.getHeaders
+                .find(hdr => CommonHttpHeader.CONTENT_TYPE.matches(hdr.getName))
+                .map(_.getElements.head.getParameterByName("charset"))
+                .flatMap(Option.apply)
+                .map(_.getValue)
+                .getOrElse(StandardCharsets.ISO_8859_1.name())
+              val jsonResponse = Source.fromInputStream(serviceClientResponse.getData, responseEncoding).getLines.mkString
+              val json = Json.parse(jsonResponse)
+              (json \ "access" \ "token" \ "id").as[String]
+            } recover {
+              case f: Exception =>
+                throw GenericIdentityException("Token could not be parsed from response from Identity", f)
+            }
+          case SC_REQUEST_ENTITY_TOO_LARGE | SC_TOO_MANY_REQUESTS =>
+            Failure(OverLimitException(buildRetryValue(serviceClientResponse), "Rate limited when getting token"))
+          case statusCode =>
+            Failure(UnexpectedStatusCodeException(statusCode, "Unexpected response from Identity"))
+        }
+      case Failure(f) =>
+        Failure(GenericIdentityException("Failure communicating with Identity when getting token", f))
```

Comment:
I assume you are handling all these exceptions further down, but if not, you need to be.